### PR TITLE
Native nvq++ driver

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

- Build a driver which contains a clang-driver.

- Define CLI args of nvq++, the rest will be forwarded to clang driver, which will build the `Jobs` (e.g., assemble, link).

- In MLIR mode: inject our dual-action frontend action (one of which is the bridge). Extend the implementation to handle MLIR, LLVM or object code emission: porting the code from `cudaq-quake`,  `cudaq-opt` and `cudaq-translate` into the frontend action.

- Various args injections (e.g., link flags, include paths) following the same logic as the existing script.

- Target support: reusing the static data inside the `.config` files as much as possible; bash script logic is converted an args parsing function associated with the Target. **[Open to comments/recommendations]**.

### TODO's:

- Porting the rest of the existing nvq++ script (some options are not ported yet)

- Refactor the core logic of  `cudaq-quake`, `cudaq-opt` and `cudaq-translate` into shared code so that we can reuse them for those apps as well as this driver (the frontend action implementation contains lots of code duplications from those tools).

### Initial runtime performance comparison

Measure the compile time (since link time would be identical) in MLIR mode. Compile time improvement ~20%.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/THIENN~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/THIENN~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

  | ./nvq++   --enable-mlir -c | ./nvq++-v1   --enable-mlir -c
-- | -- | --
static_kernel.cpp | 0m1.908s | 0m1.486s
amplitude_estimation.cpp | 0m2.182s | 0m1.663s
grover.cpp | 0m1.984s | 0m1.560s

</body>

</html>

### CODE REVIEW

**Comment by schweitzpgi** [comment]
Is there some sort of design document and discussion for this? The comments don't mention `cudaq-quake`, which would seem to be the core piece of this work. Or maybe not?

**Comment by bettinaheim** [comment]
> > Is there some sort of design document and discussion for this? The comments don't mention `cudaq-quake`, which would seem to be the core piece of this work. Or maybe not?
> 
> I don't have a design document written yet, happy to write one. Regarding the comments: indeed, I missed `cudaq-quake` for some reason :), fixed it.

@1tnguyen Yes, please create a design document in the form of a GitHub issue (and RFC).

File: tools/nvqpp-v1/nvqpp_driver.cpp [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+        clOptions.getLastArgValue(cudaq::nvqpp::options::OPT_target, "").str();
+    auto target = TargetRegistry::lookupTarget(clTargetName);
+    if (target.has_value()) {
+      targetName = clTargetName;
+      targetPlatformExtraArgs =
+          target->generateConfig(clOptions, cudaqTargetsPath);
+
+      // Insert backend config to the list of files to compile
+      const std::string backendConfigCppFile =
+          cudaqTargetsPath + "/backendConfig.cpp";
```
Comment (amccaskey) [inline tools/nvqpp-v1/nvqpp_driver.cpp:70]: I do recall that Clang / LLVM had a FileSystem abstraction, and in the past I've used one of its specializations to write a "file" to the current program memory space, without ever needing to actually output a file to disk. I wonder if we could use that here and avoid having the `backendConfig.cpp` file, and just use a string template defined in this file.

File: tools/nvqpp-v1/nvqpp_targets.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include <iostream>
+#include <regex>
+
+namespace cudaq {
+TargetPlatformConfig
+Target::generateConfig(const llvm::opt::InputArgList &clOptions,
+                       const std::filesystem::path &targetConfigSearchPath) {
+  TargetPlatformConfig config;
+  // Load from file:
+  const std::string fileName = name + std::string(".config");
```
Comment (amccaskey) [inline tools/nvqpp-v1/nvqpp_targets.cpp:20]: I would like us to figure out how to drop the bash config files and pick up a json or yaml file instead.

Comment (amccaskey) [reply tools/nvqpp-v1/nvqpp_targets.cpp:20]: The difficult thing is how to handle things like `--quantinuum-url`. Maybe the config json file can contribute extra command line arguments that can be picked up and exposed to the driver?

**Comment by bettinaheim** [comment]
@1tnguyen Actually, there is an [RFC](https://github.com/NVIDIA/cuda-quantum/issues/505) from @schweitzpgi. Let's first drive that to a conclusion before doing further implementation or prototyping work. As part of that, I would like to get to a clear list of scenarios we want to support with this change.

**Comment by 1tnguyen** [comment]
Agreed. I'll just write up a design document referring to the above [RFC](https://github.com/NVIDIA/cuda-quantum/issues/505) to facilitate further discussion. The main purpose of this draft was to experiment with LLVM/Clang to see whether we could implement `nvq++` functionalities without modifying LLVM itself.

File: tools/nvqpp-v1/nvqpp_targets.cpp (hunk 2 above)
Comment (schweitzpgi) [reply tools/nvqpp-v1/nvqpp_targets.cpp:20]: What about config files? https://releases.llvm.org/6.0.0/tools/clang/docs/UsersManual.html#configuration-files

Comment (1tnguyen) [reply tools/nvqpp-v1/nvqpp_targets.cpp:20]: +1 for LLVM config file :-) 

The difficulty of handling things like `--quantinuum-url` is that we need to make them extensible (new backends can contribute new arguments easily) while also encoding the logic to convert them into a proper `#define` string. 

Looking at Clang/LLVM, it has a combined `options.td` which virtually contains everything. Not sure if this is appropriate for our case, e.g., require new backend implementations to add entries to the argument list.

Comment (schweitzpgi) [reply tools/nvqpp-v1/nvqpp_targets.cpp:20]: We should make a list of what we suspect backends may need for support (based on our experience so far).

**Pull request closed without merging**